### ISSUE TITLE
Slow subsetting with the default number of threads

### ISSUE THREAD

**Issue by khotilov** [Rdatatable/data.table#1883]
This is related to #1660, but that PR was closed for awhile.

After installing the development version of data.table one week ago, I've noticed that subsetting have slowed down significantly, especially when running on a 16-core Linux server. I've run some experiments with different numbers of rows and DT-threads, and the results are shown below. With the default maxed-out # of OMP threads, the Linux server was swamped running at over 3100% CPU even for small tables, and was clearly not the best choice on Windows (at least for the considered data sizes and the kind of subsetting).

I've ended up just using either two or three threads. But having to remember to set it every time, or to re-set it each time after RStudio server restores my session is a bit annoying. Perhaps, some adaptive assignment of the default number of threads might make sense? E.g., something like max(1, floor(log10(nrows)) - 2) and no more than the number of physical cores?

Running on a 16 cores / 32 threads Linux server (nothing else was running at the time):

``` R
> library(data.table)
data.table 1.9.7 IN DEVELOPMENT built 2016-10-11 05:48:49 UTC; travis
  The fastest way to learn (by data.table authors): https://www.datacamp.com/courses/data-analysis-the-data-table-way
  Documentation: ?data.table, example(data.table) and browseVignettes("data.table")
  Release notes, videos and slides: http://r-datatable.com
> library(microbenchmark)
>
> getDTthreads()
[1] 32
>
> f <- function(n) { setDTthreads(n); tmp[sel,]}
>
> tmp <- as.data.table(matrix(rnorm(10000*10), ncol=10))
> sel <- as.logical(rbinom(10000, 1, 0.75))
> microbenchmark(f(1), f(2), f(4), f(8), f(16), f(24), f(31), f(32), times=5)
Unit: microseconds
  expr        min         lq        mean     median         uq        max neval cld
  f(1)    805.986    809.302    959.3076    818.821   1055.329   1307.100     5  a
  f(2)    667.242    777.297   2753.4888    782.712    984.754  10555.439     5  a
  f(4)    693.011    704.204    787.8422    787.633    869.192    885.171     5  a
  f(8)    832.505    851.885    886.9588    863.308    890.116    996.980     5  a
 f(16)   1220.499   1262.749   2847.4750   1317.784   1787.721   8648.622     5  a
 f(24)   1162.788   1167.810   6198.9242   1288.923   1543.947  25831.153     5  a
 f(31)   1093.950   1272.632  65441.6472   1460.030   1670.858 321710.766     5  a
 f(32) 877883.448 878172.905 893367.5322 886586.371 905078.450 919116.487     5   b
>
> tmp <- as.data.table(matrix(rnorm(100000*10), ncol=10))
> sel <- as.logical(rbinom(100000, 1, 0.75))
> microbenchmark(f(1), f(2), f(4), f(8), f(16), f(24), f(31), f(32), times=5)
Unit: milliseconds
  expr        min         lq       mean     median         uq        max neval cld
  f(1)   4.002226   4.092673   4.617931   4.193665   5.382385   5.418707     5  a
  f(2)   2.506109   3.149362   3.303735   3.259054   3.282128   4.322021     5  a
  f(4)   1.970885   2.261464   2.533936   2.444877   2.505059   3.487393     5  a
  f(8)   1.924474   2.018732   2.138352   2.125057   2.259300   2.364197     5  a
 f(16)   2.254166   2.782224   4.665037   3.051369   3.887746  11.349681     5  a
 f(24)   2.357963   2.392597   4.309452   2.486199   2.747933  11.562570     5  a
 f(31)   2.193412   2.602528  11.311535   5.004872  20.173821  26.583042     5  a
 f(32) 888.226689 891.343169 906.466687 894.008800 907.523427 951.231349     5   b
>
> tmp <- as.data.table(matrix(rnorm(1000000*10), ncol=10))
> sel <- as.logical(rbinom(1000000, 1, 0.75))
> microbenchmark(f(1), f(2), f(4), f(8), f(16), f(24), f(31), f(32), times=5)
Unit: milliseconds
  expr       min        lq      mean    median        uq       max neval cld
  f(1)  51.71999  59.94565 103.66509  86.81867 147.57261 172.26855     5  a
  f(2)  23.61526  31.56658  91.16605  42.83279 147.75449 210.06113     5  a
  f(4)  17.04975  17.24534  20.97919  21.47207  24.10658  25.02223     5  a
  f(8)  15.62397  17.01034  23.06007  22.04185  25.13124  35.49296     5  a
 f(16)  15.64664  18.34823  83.73781  22.19100 170.49213 192.01107     5  a
 f(24)  22.77587  34.00246 130.85567 169.87868 196.96246 230.65890     5  a
 f(31)  22.90744  23.12211  61.85601  23.51561  51.16077 188.57411     5  a
 f(32) 862.84840 891.22982 917.84761 908.24769 962.51034 964.40182     5   b
>
> tmp <- as.data.table(matrix(rnorm(10000000*10), ncol=10))
> sel <- as.logical(rbinom(10000000, 1, 0.75))
> microbenchmark(f(1), f(2), f(4), f(8), f(16), f(24), f(31), f(32), times=5)
Unit: milliseconds
  expr       min        lq      mean    median        uq       max neval cld
  f(1)  673.8642  701.2184  771.2369  748.5801  819.5180  913.0039     5 a
  f(2)  528.6492  575.7349  673.9835  605.5423  790.7594  869.2315     5 a
  f(4)  434.3270  544.6065  602.3636  570.7178  712.1214  750.0454     5 a
  f(8)  589.1129  606.2126  693.5335  656.7463  786.0603  829.5352     5 a
 f(16)  354.7724  591.1855  650.5706  671.9307  737.6038  897.3605     5 a
 f(24)  678.2691  767.7509  930.9970  901.5826 1050.7196 1256.6630     5 ab
 f(31)  892.8193 1313.6712 1338.2678 1330.9336 1496.1389 1657.7761     5  bc
 f(32) 1204.2505 1381.4561 1563.4297 1515.7371 1713.8750 2001.8297     5   c
```

On a 4 cores / 8 threads Windows laptop:

``` R
> library(data.table)
data.table 1.9.7 IN DEVELOPMENT built 2016-10-11 05:48:49 UTC; travis
  The fastest way to learn (by data.table authors): https://www.datacamp.com/courses/data-analysis-the-data-table-way
  Documentation: ?data.table, example(data.table) and browseVignettes("data.table")
  Release notes, videos and slides: http://r-datatable.com
> library(microbenchmark)
> options(width=180)
> 
> getDTthreads()
[1] 8
> 
> f <- function(n) { setDTthreads(n); tmp[sel,]}
> 
> tmp <- as.data.table(matrix(rnorm(10000*10), ncol=10))
> sel <- as.logical(rbinom(10000, 1, 0.75))
> microbenchmark(f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8), times=5)
Unit: microseconds
 expr      min       lq      mean   median       uq      max neval cld
 f(1)  861.249  883.058  893.4924  899.736  907.861  915.558     5 a  
 f(2) 1447.530 1456.938 1502.1818 1479.603 1492.860 1633.978     5 ab 
 f(3) 1608.748 1614.735 1702.0566 1730.622 1744.306 1811.872     5 ab 
 f(4) 2048.779 2152.266 2268.4106 2306.213 2315.194 2519.601     5 ab 
 f(5) 2279.272 2311.345 2336.0620 2333.154 2369.503 2387.036     5 ab 
 f(6) 2598.713 2632.923 2767.8834 2688.515 2949.798 2969.468     5  b 
 f(7) 2926.278 3047.298 3066.5408 3050.291 3063.975 3244.862     5  bc
 f(8) 3407.362 3627.593 4706.5902 3654.105 3772.987 9070.904     5   c
> 
> tmp <- as.data.table(matrix(rnorm(100000*10), ncol=10))
> sel <- as.logical(rbinom(100000, 1, 0.75))
> microbenchmark(f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8), times=5)
Unit: milliseconds
 expr      min       lq     mean   median       uq      max neval cld
 f(1) 3.280356 3.317133 3.799585 3.551474 3.805486 5.043478     5  a 
 f(2) 3.315850 3.488612 3.636059 3.690882 3.789237 3.895716     5  a 
 f(3) 3.386408 4.265189 4.630728 4.507657 5.305615 5.688772     5  ab
 f(4) 3.832855 4.033841 4.706334 4.638511 4.981044 6.045417     5  ab
 f(5) 3.859796 4.174104 5.085813 4.462755 6.171140 6.761271     5  ab
 f(6) 4.437098 4.682557 4.807426 4.685551 4.889531 5.342392     5  ab
 f(7) 4.634663 4.689400 4.850873 4.698380 5.087952 5.143971     5  ab
 f(8) 4.789465 5.020386 5.401063 5.173905 5.588707 6.432850     5   b
> 
> tmp <- as.data.table(matrix(rnorm(1000000*10), ncol=10))
> sel <- as.logical(rbinom(1000000, 1, 0.75))
> microbenchmark(f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8), times=5)
Unit: milliseconds
 expr      min       lq     mean    median        uq       max neval cld
 f(1) 38.40637 43.11288 43.32379  44.19436  44.99531  45.91001     5   a
 f(2) 26.03416 31.40050 31.82060  32.37678  33.89572  35.39585     5   a
 f(3) 27.53258 31.01863 43.88424  31.39537  32.18478  97.28985     5   a
 f(4) 25.94350 25.96660 30.44526  32.82579  33.28764  34.20276     5   a
 f(5) 27.40600 33.95260 61.49442  36.38368 103.63717 106.09263     5   a
 f(6) 25.91827 34.06592 74.99215 104.72506 105.11036 105.14115     5   a
 f(7) 26.81288 33.48563 60.66669  33.53395 104.37954 105.12148     5   a
 f(8) 25.33969 27.09640 44.29459  27.79343  34.17112 107.07233     5   a
> 
> tmp <- as.data.table(matrix(rnorm(10000000*10), ncol=10))
> sel <- as.logical(rbinom(10000000, 1, 0.75))
> microbenchmark(f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8), times=5)
Unit: milliseconds
 expr      min       lq     mean   median       uq      max neval cld
 f(1) 487.1292 491.9212 497.6124 491.9811 504.4007 512.6296     5   c
 f(2) 369.1809 370.3642 386.2093 375.3516 378.7115 437.4380     5  b 
 f(3) 345.3490 346.4425 352.1456 353.4633 355.4916 359.9817     5 ab 
 f(4) 345.9293 350.2689 350.7956 350.7868 351.6900 355.3030     5 ab 
 f(5) 230.1949 277.3694 304.6530 277.4032 360.5787 377.7186     5 a  
 f(6) 353.1892 355.9898 364.7204 360.1216 373.7424 380.5589     5  b 
 f(7) 349.7866 353.1272 358.8474 359.7692 364.6677 366.8862     5  b 
 f(8) 347.9183 351.9799 353.4224 352.9938 356.1005 358.1194     5 ab
```

Session infos:

```
R version 3.3.1 (2016-06-21)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server release 6.7 (Santiago)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] microbenchmark_1.4-2.1 data.table_1.9.7

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.5      codetools_0.2-14 lattice_0.20-33  mvtnorm_1.0-5    zoo_1.7-13
 [6] MASS_7.3-45      grid_3.3.1       plyr_1.8.4       gtable_0.2.0     scales_0.4.0
[11] ggplot2_2.1.0    multcomp_1.4-6   Matrix_1.2-6     sandwich_2.3-4   TH.data_1.0-7
[16] splines_3.3.1    munsell_0.4.3    survival_2.39-5  colorspace_1.2-6
```

```
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] microbenchmark_1.4-2.1 data.table_1.9.7      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.7      codetools_0.2-15 lattice_0.20-34  mvtnorm_1.0-5    zoo_1.7-13      
 [6] MASS_7.3-45      grid_3.3.0       plyr_1.8.4       gtable_0.2.0     scales_0.4.0    
[11] ggplot2_2.1.0    multcomp_1.4-6   Matrix_1.2-7.1   sandwich_2.3-4   TH.data_1.0-7   
[16] splines_3.3.0    munsell_0.4.3    survival_2.39-5  colorspace_1.2-6
```

**Comment by mattdowle** [comment]
Many thanks.  Will take a look.

**Comment by andrewrech** [comment]
I use a similar Linux server and see the same slowdown using `DT_new <- copy(DT)` which with a large DT completes in 4 seconds with 8 threads and fails to complete with threads == cores == 32.

**Comment by mattdowle** [comment]
@khotilov I reproduced your results, thanks.

@andrewrech Are you sure that it's `copy(DT)` to blame and not something else around that call to `copy(DT)`.  No parallelism has been applied to `copy()` yet as far as I can make out.  It just calls the single threaded base R duplicate() at C level.  That size 12000*25000, if all 8 byte numeric, works out at about 2.2GB,  so should be fine.  Worth trying again with a fresh R session, a fresh install of data.table dev (in case other crash bugs are now fixed) to see if it really is `copy(DT)` that caused it?
